### ISSUE TITLE
How to edit a car of the scene

### ISSUE THREAD

**Issue by jelleopard** [OPEN-AIR-SUN/mars#35]
> For example, you can edit objects in a scene by changing the pose and rotation of a car. Concretely, the last dimension of rotation represents `object_id`, the last 2 dimensions of pose represent `object_id` and `axis`. ![image](https://user-images.githubusercontent.com/107318439/258739023-1c168b56-2f14-4b7e-b1e3-c3c221eabcf3.png)

> > 
> 
> We now don't support directly editing the car type, but it can be implemented via manually modifying the car latents. But you can change the appearance of a car to an existing one in the scene. To edit car trajectory, you can generate your own car pose and assign it to `pose` and `rotation` across the scene.

@xBeho1der 
Thanks for your detailed reply, Your algorithm models the scene and dynamic objects separately, how do I change the style of a moving car or remove the car when doing a rendering，etc.
May I ask how to change the appearance of the car to an existing one in the scene. 
Does object_id represent the currently rendered dynamic object, and if I change the id can I change the type of the dynamic object?
How to change the following code to achieve scene editing, can you give some comments or example codes

pose = batch_obj_dyn[..., :3]
rotation = batch_obj_dyn[..., 3]
pose[:, :, 0, 2] = pose[:, :, 0, 2]
rotation[:, :, 0] = rotation[:, :, 0]
batch_obj_dyn[..., :3] = pose
batch_obj_dyn[..., 3] = rotation

**Comment by zwlvd** [comment]
I am trying to edit a car as well. From my perspective, maybe the edit can be completed by change the x,y,z,yaw_angle in the obj_info. While the problem is there are many frames information where all the objects'  x,y,z,yaw_angle should be changed, it seems to be a huge work.

**Comment by jelleopard** [comment]
The reproduction result on kitti/0006 is poor.
![image](https://github.com/OPEN-AIR-SUN/mars/assets/43605327/8a4e90c5-0754-4d8f-8f9a-5e130eccd7b3)
But the config file you provided uses latent_codes_car_van_truck.pt
![image](https://github.com/OPEN-AIR-SUN/mars/assets/43605327/9949f116-5be2-4c97-af7e-a45306ecd856)

My used config as follows:
!!python/object:nerfstudio.engine.trainer.TrainerConfig
_target: !!python/name:nerfstudio.engine.trainer.Trainer ''
data: &id003 !!python/object/apply:pathlib.PosixPath
- /
- home
- mars
- data
- kitti
- training
- image_02
- '0006'
experiment_name: '0006'
load_checkpoint: null
load_config: null
load_dir: null
load_step: null
log_gradients: true
logging: !!python/object:nerfstudio.configs.base_config.LoggingConfig
  local_writer: !!python/object:nerfstudio.configs.base_config.LocalWriterConfig
    _target: !!python/name:nerfstudio.utils.writer.LocalWriter ''
    enable: true
    max_log_size: 10
    stats_to_track: !!python/tuple
    - !!python/object/apply:nerfstudio.utils.writer.EventName
      - Train Iter (time)
    - !!python/object/apply:nerfstudio.utils.writer.EventName
      - Train Rays / Sec
    - !!python/object/apply:nerfstudio.utils.writer.EventName
      - Test PSNR
    - !!python/object/apply:nerfstudio.utils.writer.EventName
      - Vis Rays / Sec
    - !!python/object/apply:nerfstudio.utils.writer.EventName
      - Test Rays / Sec
    - !!python/object/apply:nerfstudio.utils.writer.EventName
      - ETA (time)
  max_buffer_size: 20
  profiler: basic
  relative_log_dir: !!python/object/apply:pathlib.PosixPath []
  steps_per_log: 10
machine: !!python/object:nerfstudio.configs.base_config.MachineConfig
  dist_url: auto
  machine_rank: 0
  num_gpus: 1
  num_machines: 1
  seed: 42
max_num_iterations: 600000
method_name: ablation-no-depth-kitti
mixed_precision: false
optimizers:
  background_model:
    optimizer: !!python/object:nerfstudio.engine.optimizers.RAdamOptimizerConfig
      _target: &id001 !!python/name:torch.optim.radam.RAdam ''
      eps: 1.0e-15
      lr: 0.001
      max_norm: null
      weight_decay: 0
    scheduler: !!python/object:nerfstudio.engine.schedulers.ExponentialDecaySchedulerConfig
      _target: &id002 !!python/name:nerfstudio.engine.schedulers.ExponentialDecayScheduler ''
      lr_final: 1.0e-05
      lr_pre_warmup: 1.0e-08
      max_steps: 200000
      ramp: cosine
      warmup_steps: 0
  learnable_global:
    optimizer: !!python/object:nerfstudio.engine.optimizers.RAdamOptimizerConfig
      _target: *id001
      eps: 1.0e-15
      lr: 0.001
      max_norm: null
      weight_decay: 0
    scheduler: !!python/object:nerfstudio.engine.schedulers.ExponentialDecaySchedulerConfig
      _target: *id002
      lr_final: 1.0e-05
      lr_pre_warmup: 1.0e-08
      max_steps: 200000
      ramp: cosine
      warmup_steps: 0
  object_model:
    optimizer: !!python/object:nerfstudio.engine.optimizers.RAdamOptimizerConfig
      _target: *id001
      eps: 1.0e-15
      lr: 0.005
      max_norm: null
      weight_decay: 0
    scheduler: !!python/object:nerfstudio.engine.schedulers.ExponentialDecaySchedulerConfig
      _target: *id002
      lr_final: 1.0e-05
      lr_pre_warmup: 1.0e-08
      max_steps: 200000
      ramp: cosine
      warmup_steps: 0
output_dir: !!python/object/apply:pathlib.PosixPath
- outputs
pipeline: !!python/object:nsg.nsg_pipeline.NSGPipelineConfig
  _target: !!python/name:nsg.nsg_pipeline.NSGPipeline ''
  datamanager: !!python/object:nsg.data.nsg_datamanager.NSGkittiDataManagerConfig
    _target: !!python/name:nsg.data.nsg_datamanager.NSGkittiDataManager ''
    camera_optimizer: !!python/object:nerfstudio.cameras.camera_optimizers.CameraOptimizerConfig
      _target: !!python/name:nerfstudio.cameras.camera_optimizers.CameraOptimizer ''
      mode: 'off'
      optimizer: !!python/object:nerfstudio.engine.optimizers.AdamOptimizerConfig
        _target: !!python/name:torch.optim.adam.Adam ''
        eps: 1.0e-15
        lr: 0.0006
        max_norm: null
        weight_decay: 0
      orientation_noise_std: 0.0
      param_group: camera_opt
      position_noise_std: 0.0
      scheduler: !!python/object:nerfstudio.engine.schedulers.ExponentialDecaySchedulerConfig
        _target: *id002
        lr_final: null
        lr_pre_warmup: 1.0e-08
        max_steps: 10000
        ramp: cosine
        warmup_steps: 0
    camera_res_scale_factor: 1.0
    data: *id003
    dataparser: !!python/object:nsg.data.nsg_dataparser.NSGkittiDataParserConfig
      _target: !!python/name:nsg.data.nsg_dataparser.NSGkitti ''
      add_input_rows: -1
      alpha_color: white
      bckg_only: false
      box_scale: 1.5
      car_nerf_state_dict_path: !!python/object/apply:pathlib.PosixPath
      - /
      - home
      - mars
      - latents
      - KITTI-MOT
      - car-nerf-state-dict
      - epoch_670.ckpt
      car_object_latents_path: !!python/object/apply:pathlib.PosixPath
      - /
      - home
      - sunpenghui
      - mars
      - latents
      - KITTI-MOT
      - car-object-latents
      - latent_codes06.pt
      chunk: 32768
      data: !!python/object/apply:pathlib.PosixPath
      - data
      - kitti
      - training
      - image_02
      - '0006'
      dataset_type: kitti
      far_plane: 150.0
      first_frame: 5
      last_frame: 260
      max_input_objects: -1
      near_plane: 0.5
      netchunk: 65536
      novel_view: left
      obj_only: false
      obj_opaque: true
      object_setting: 0
      render_only: false
      scale_factor: 1
      scene_scale: 1.0
      semantic_mask_classes: []
      semantic_path: !!python/object/apply:pathlib.PosixPath []
      split_setting: reconstruction
      use_car_latents: true
      use_depth: false
      use_obj: true
      use_object_properties: true
      use_semantic: false
    eval_image_indices: !!python/tuple
    - 0
    eval_num_images_to_sample_from: -1
    eval_num_rays_per_batch: 4096
    eval_num_times_to_repeat_images: -1
    patch_size: 1
    train_num_images_to_sample_from: -1
    train_num_rays_per_batch: 4096
    train_num_times_to_repeat_images: -1
  model: !!python/object:nsg.models.scene_graph.SceneGraphModelConfig
    _target: !!python/name:nsg.models.scene_graph.SceneGraphModel ''
    background_color: last_sample
    background_model: !!python/object:nsg.models.nerfacto.NerfactoModelConfig
      _target: !!python/name:nsg.models.nerfacto.NerfactoModel ''
      background_color: black
      collider_params:
        far_plane: 6.0
        near_plane: 2.0
      disable_scene_contraction: false
      distortion_loss_mult: 0.002
      enable_collider: true
      eval_num_rays_per_chunk: 4096
      far_plane: 150.0
      hidden_dim: 64
      hidden_dim_color: 64
      hidden_dim_transient: 64
      log2_hashmap_size: 19
      loss_coefficients:
        rgb_loss_coarse: 1.0
        rgb_loss_fine: 1.0
      max_res: 2048
      near_plane: 0.05
      num_coarse_samples: 24
      num_levels: 16
      num_nerf_samples_per_ray: 97
      num_proposal_iterations: 2
      num_proposal_samples_per_ray: !!python/tuple
      - 256
      - 128
      obj_feat_dim: 0
      orientation_loss_mult: 0.0001
      pred_normal_loss_mult: 0.001
      predict_normals: false
      proposal_net_args_list:
      - hidden_dim: 16
        log2_hashmap_size: 17
        max_res: 128
        num_levels: 5
        use_linear: false
      - hidden_dim: 16
        log2_hashmap_size: 17
        max_res: 256
        num_levels: 5
        use_linear: false
      proposal_update_every: 5
      proposal_warmup: 5000
      proposal_weights_anneal_max_num_iters: 1000
      proposal_weights_anneal_slope: 10.0
      sampler: proposal
      use_average_appearance_embedding: true
      use_gradient_scaling: false
      use_proposal_weight_anneal: true
      use_same_proposal_network: false
      use_single_jitter: true
    collider_params:
      far_plane: 6.0
      near_plane: 2.0
    debug_object_pose: false
    depth_loss_mult: 0.01
    depth_loss_type: !!python/object/apply:nerfstudio.model_components.losses.DepthLossType
    - 1
    depth_sigma: 0.05
    enable_collider: true
    eval_num_rays_per_chunk: 4096
    far_plane: 1000.0
    interlevel_loss_mult: 1.0
    is_euclidean_depth: false
    latent_size: 256
    loss_coefficients:
      rgb_loss_coarse: 1.0
      rgb_loss_fine: 1.0
    max_num_obj: -1
    mono_depth_loss_mult: 0.0
    near_plane: 0.05
    object_model_template: !!python/object:nsg.models.car_nerf.CarNeRFModelConfig
      _target: !!python/name:nsg.models.car_nerf.CarNeRF ''
      background_color: black
      collider_params:
        far_plane: 6.0
        near_plane: 2.0
      enable_collider: true
      eval_num_rays_per_chunk: 4096
      loss_coefficients:
        rgb_loss_coarse: 1.0
        rgb_loss_fine: 1.0
      num_coarse_samples: 32
      num_fine_samples: 97
      optimize_latents: false
    object_ray_sample_strategy: remove-bg
    object_representation: class-wise
    object_warmup_steps: 1000
    orientation_loss_mult: 0.0001
    pred_normal_loss_mult: 0.001
    predict_normals: false
    ray_add_input_rows: -1
    semantic_loss_mult: 1.0
    should_decay_sigma: false
    sigma_decay_rate: 0.9998
    sky_model: !!python/object:nsg.models.sky_model.SkyModelConfig
      _target: !!python/name:nsg.models.sky_model.SkyModel ''
      collider_params:
        far_plane: 6.0
        near_plane: 2.0
      enable_collider: true
      eval_num_rays_per_chunk: 4096
      hidden_dim: 128
      loss_coefficients:
        rgb_loss_coarse: 1.0
        rgb_loss_fine: 1.0
      num_layers: 5
    starting_depth_sigma: 4.0
    use_interlevel_loss: true
    use_sky_model: false
project_name: nerfstudio-project
relative_model_dir: !!python/object/apply:pathlib.PosixPath
- nerfstudio_models
save_only_latest_checkpoint: false
steps_per_eval_all_images: 300000
steps_per_eval_batch: 500
steps_per_eval_image: 1000
steps_per_save: 5000
timestamp: 2023-08-08_140643
use_grad_scaler: true
viewer: !!python/object:nerfstudio.configs.base_config.ViewerConfig
  image_format: jpeg
  jpeg_quality: 90
  max_num_display_images: 512
  num_rays_per_chunk: 32768
  quit_on_train_completion: false
  relative_log_filename: viewer_log_filename.txt
  websocket_host: 0.0.0.0
  websocket_port: null
  websocket_port_default: 7007
vis: wandb

**Comment by jelleopard** [comment]
> I am trying to edit a car as well. From my perspective, maybe the edit can be completed by change the x,y,z,yaw_angle in the obj_info. While the problem is there are many frames information where all the objects' x,y,z,yaw_angle should be changed, it seems to be a huge work.

Thanks for your reply. Can you provide more details about follow codes
pose = batch_obj_dyn[..., :3]
rotation = batch_obj_dyn[..., 3]
pose[:, :, 0, 2] = pose[:, :, 0, 2]
rotation[:, :, 0] = rotation[:, :, 0]

**Comment by zwlvd** [comment]
> > I am trying to edit a car as well. From my perspective, maybe the edit can be completed by change the x,y,z,yaw_angle in the obj_info. While the problem is there are many frames information where all the objects' x,y,z,yaw_angle should be changed, it seems to be a huge work.
> 
> Thanks for your reply. Can you provide more details about follow codes pose = batch_obj_dyn[..., :3] rotation = batch_obj_dyn[..., 3] pose[:, :, 0, 2] = pose[:, :, 0, 2] rotation[:, :, 0] = rotation[:, :, 0]

Well, I haven't made such try for I think my idea is not so good . Maybe the author will provide some more useful suggestions.

**Comment by wuzirui** [comment]
Hi, thank you for your interest in our project. However, I may need to mention that you are not supposed to open multiple issues for the same question. If the responses in previous issues are not addressed, please click the `Reopen` button. Otherwise, your GitHub account might be prohibited from creating new issues.

Duplicate with #28.

**Comment by xBeho1der** [comment]
> > I am trying to edit a car as well. From my perspective, maybe the edit can be completed by change the x,y,z,yaw_angle in the obj_info. While the problem is there are many frames information where all the objects' x,y,z,yaw_angle should be changed, it seems to be a huge work.
> 
> Thanks for your reply. Can you provide more details about follow codes pose = batch_obj_dyn[..., :3] rotation = batch_obj_dyn[..., 3] pose[:, :, 0, 2] = pose[:, :, 0, 2] rotation[:, :, 0] = rotation[:, :, 0]

Actually, the first two lines mean reading the object pose and rotation information from the batch. You can edit the pose or rotation by adding and subtracting in the last two lines.

**Comment by jelleopard** [comment]
> 

Thanks for your reminder, I also tried to reopen the problem, but I did not find the reopen button, because the problem was not closed by me, and I found that there is no other better solution than to recreate the problem. I'm sorry for the inconvenience.
